### PR TITLE
Feature/380 nextcloud integration chat utils

### PR DESCRIPTION
**Pull request by absurd-turtle**
This PR introduces a chat component which serves as an example on how to interact with the Nextcloud Talk API.

Not in scope of this PR:
 - translations: this will be done when the chat is integrated in the landing page or Photo Layer
 - fancy UI: this UI is there to test the chat integration including the polling mechanism 

## Basics

- [x] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

File: frontend/src/features/nextcloud_integration/api/chat.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  //When a user is deleted from the server or removed from all their conversations, 1 "One to one" rooms are converted to this type)
+  FormerOneToOne,
+}
+export type CreateConversationOptions = {
+  roomType: ConversationType;
+  // User ID (roomType = 1), group ID (roomType = 2 - optional), circle ID (roomType = 2, source = 'circles'], only available with circles-support capability)
+  invite?: string;
+  // The source for the invite, only supported on roomType = 2 for groups and circles (only available with circles-support capability)
+  source?: string;
+  // Conversation name up to 255 characters (Not available for roomType = 1)
```
Comment (markus2330) [inline frontend/src/features/nextcloud_integration/api/chat.ts:97]: Why not use the symbolic name? Especially as there seem to be an undocumented off-by-one.

```suggestion
  // Conversation name up to 255 characters (not available for roomType = OneToOne)
```

File: frontend/src/features/nextcloud_integration/api/chat.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  roomName?: string;
+  // Type of an object this room references, currently only allowed value is 'room' to indicate the parent of a breakout room (See Object types)
```
Comment (markus2330) [inline frontend/src/features/nextcloud_integration/api/chat.ts:99]: Where is object types?

```suggestion
  // Type of an object this room references, currently only allowed value is 'room' to indicate the parent of a breakout room (see object types)
```

File: frontend/src/features/nextcloud_integration/api/chat.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  objectType?: string;
+  // ID of an object this room references, room token is used for the parent of a breakout room
+  objectId?: string;
+};
+export const createConversation = async (options: CreateConversationOptions) => {
+  const http = createNextcloudAPI();
+  http.defaults.headers['OCS-APIRequest'] = true;
+  return http.post('/ocs/v2.php/apps/spreed/api/v4/room', {
+    ...options,
+    roomType: options.roomType + 1,
```
Comment (markus2330) [inline frontend/src/features/nextcloud_integration/api/chat.ts:109]: Isn't it better to fix this in the enum?

File: frontend/src/features/nextcloud_integration/api/chat.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  timestamp: number;
+  // Empty for normal chat message or the type of the system message (untranslated)
+  systemMessage: string;
+  // Currently known types are comment, comment_deleted, system, and command
+  messageType: string;
+  // True if the user can post a reply to this message (only available with chat-replies capability)
+  isReplyable: boolean;
+  // A reference string that was given while posting the message to be able to identify a sent message again (only available with chat-reference-id capability)
+  referenceId: string;
+  // Message string with placeholders (see Rich Object String)
```
Comment (markus2330) [inline frontend/src/features/nextcloud_integration/api/chat.ts:164]: I cannot find a "Rich Object String"

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you for the update. Please fix CI here, too.

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
> The prototype chat component is available for testing at /chat at the moment.
> Should I leave the route for testing purposes or remove it so it is not user facing?

Yes, leave it :+1: 

But I couldn't get it running, I constantly get `401` errors `xhr.js:247  GET https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room 401` even without pressing anything.

Furthermore, I cannot press the "Submit" button to create a conversation.

```
react-dom.development.js:86 Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component.

{roomType: 0, objectType: 'room', invite: 'yvonne'}
xhr.js:247 
 POST https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room 401
```

File: frontend/src/features/nextcloud_integration/api/chat.ts (hunk 2 above)
Comment (absurd-turtle) [reply frontend/src/features/nextcloud_integration/api/chat.ts:99]: Most of the type documentation I took directly from the Nextcloud API docs.
I will add a link to the docs.

**Comment by absurd-turtle** [comment]
I'll see what I can do to clean up the issues.
I don't want to spend too much time though as this will not be used in the near future anyways.
Especially the prototype chat component.

> > The prototype chat component is available for testing at /chat at the moment.
> > Should I leave the route for testing purposes or remove it so it is not user facing?
> 
> Yes, leave it 👍
> 
> But I couldn't get it running, I constantly get `401` errors `xhr.js:247 GET https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room 401` even without pressing anything.
> 
> Furthermore, I cannot press the "Submit" button to create a conversation.
> 
> ```
> react-dom.development.js:86 Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component.
> 
> {roomType: 0, objectType: 'room', invite: 'yvonne'}
> xhr.js:247 
>  POST https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room 401
> ```

The error is probably due to CORS as we apparently haven't handled these calls yet.
Please try it with `chromium --disable-web-security`

The calls that are failing are polling for available conversations.

**Comment by markus2330** [comment]
> Please try it with chromium --disable-web-security

I tried again with chromium Version 116.0.5845.96, of course with --disable-web-security:
```
Access to XMLHttpRequest at 'https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room' from origin 'http://localhost:5173' has been blocked by CORS policy: Request header field ocs-apirequest is not allowed by Access-Control-Allow-Headers in preflight response.
AxiosError {message: 'Network Error', name: 'AxiosError', code: 'ERR_NETWORK', config: {…}, request: XMLHttpRequest, …}
xhr.js:247 
 GET https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room net::ERR_FAILED
```

When I simply visit https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room I get "Zugriff verboten
CSRF check failed".

Can you write a short test protocol that it is clear what worked for you?

**Comment by absurd-turtle** [comment]
> > Please try it with chromium --disable-web-security
> 
> I tried again with chromium Version 116.0.5845.96, of course with --disable-web-security:
> 
> ```
> Access to XMLHttpRequest at 'https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room' from origin 'http://localhost:5173' has been blocked by CORS policy: Request header field ocs-apirequest is not allowed by Access-Control-Allow-Headers in preflight response.
> AxiosError {message: 'Network Error', name: 'AxiosError', code: 'ERR_NETWORK', config: {…}, request: XMLHttpRequest, …}
> xhr.js:247 
>  GET https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room net::ERR_FAILED
> ```
> 
> When I simply visit https://cloud.permaplant.net/ocs/v2.php/apps/spreed/api/v4/room I get "Zugriff verboten CSRF check failed".
> 
> Can you write a short test protocol that it is clear what worked for you?

Ok that's good.
Simply visiting won't work as the request has to be authenticated.

## TC-018 - Chat: Create conversation
- Description: A conversation can be created at /chat
- Preconditions:
  - User must be logged in.
- Test Steps:
  1. Navigate to /chat and click create conversation
  2. Fill out the Invite with a valid Nextcloud user id.
  3. Type a name of choice in the Room Name field.
  4. Click submit.
- Expected Result:
  - The conversation is created in Nextcloud.
  - The conversation will appear in the conversation list after some time.
- Actual Result:
- Test Result:
- Notes: Currently only works with CORS disabled.

## TC-019 - Chat: Send message
- Description: Send a message to a Nextcloud conversation.
- Preconditions:
  - User must be logged in.
  - A conversation for the user must exist.
- Test Steps:
  1. Navigate to /chat and select an available conversation
  2. Type a message in the message field.
  3. Click Send.
- Expected Result:
  - The message is sent to the chat and appears in Nextcloud.
  - A success message is displayed.
  - The message will appear in the conversation history after at most 5 seconds.
- Actual Result:
- Test Result:
- Notes: Currently only works with CORS disabled.

**Comment by markus2330** [comment]
Thank you, :heart: 

Will later test on dev.permaplan.net

**Pull request merged**